[Feat/#1] 개발환경 셋업 및 컨벤션 적용

derrickkim0109 opened the pull request:
## PR 타입
어떤 변경에 대한 PR인가요?

<!-- 아래 체크리스트 중 해당 되는 부분에 체크해주세요 "x". -->

- [ ] 버그 수정
- [ ] Feature / 신기능
- [ ] 코드 스타일 업데이트 (formatting, local variables)
- [ ] 리팩토링 (no functional changes, no api changes)
- [ ] 빌드 관련 변경사항
- [ ] CI 관련된 변경사항
- [ ] Documentation content changes
- [ ] 구조 변경
- [x] Other... 추가 설명 요망:
    - tuist 초기 설정
    - PR, Commit Template 설정

## 배경

<!-- 해당 PR에서 개발한 내용에 대해 알아야 하는 배경지식을 작성해주세요. -->

- Micro Architecture 개념을 토대로 Root 모듈을 설정한다. Root 모듈 별로 Interface, Testing, Tests, Example을 생성할 수 있어야 한다.
- PR, Commit Template을 토대로 내용을 작성할 수 있어야 한다. 

## 목표

<!-- 해당 PR에서 개발/수정 한 기능의 목표를 작성해주세요. -->

1. 각 모듈 별(Feature, Serive, Network, Shared) Sample 객체 생성 
2. make module 명령어를 활용해 Root 모듈 별 하위 모듈을 생성한다.
3. PR, Commit Template을 토대로 내용을 작성한다.

## 결과 (Optional)

<!-- 개발한 내용의 결과를 설명해주세요. !-->

1. 기존에 Utility 모듈을 Foundation으로 네이밍 하기로 했으나 Swift의 Foundation 프레임워크와 네이밍이 겹치는 부분이 있어 Shared로 변경 (Foundation -> Shared)

2. make module

<img width="648" alt="스크린샷 2024-04-01 14 38 00" src="https://github.com/project-lyrics/app-iOS/assets/59466342/e90d92f2-db93-4b40-b1bb-699fb86fd6c9">

1. Terminal 내에서 make module 명령어 입력 
2. 원하는 모듈 입력 (대문자, 소문자 가능 - 명령어 예시: core)

<img width="658" alt="스크린샷 2024-04-01 14 41 30" src="https://github.com/project-lyrics/app-iOS/assets/59466342/b3b38343-5502-4bfd-a825-8b4cf4a22206">

> 내가 만들 모듈 이름(User -> CoreNetworking으로 생성 됨

<img width="647" alt="스크린샷 2024-04-01 14 41 55" src="https://github.com/project-lyrics/app-iOS/assets/59466342/3c7b0fd4-af35-4209-93f9-8f495feb069f">

> Tests, Testing 모듈 y,n 여부로 생성

3. PR, Commit 생성 시 Template을 확인할 수 있다. 
- PR을 생성하면 PR Template을 확인할 수 있따.
- Terminal에서 git commit 명령어 입력시 Template을 확인할 수 있다. 
  - Commit Template 적용 명령어: **git config commit.template ~/MyProject/.github/commit-template.txt**

<!-- 
gif, 사진을 위한 Table

| | | |
|--|--|--|
|<img src="" width="250">|<img src="" width="250">|<img src="" width="250">|
-->

## Trouble or Trouble Shooting(Optional)

<!-- 해당 PR에서 발생한 문제를 해결한 내용을 작성해주세요.  !-->

derrickkim0109 committed:
add: Module 객체 추가

derrickkim0109 committed:
add: Project Environment 객체 추가

derrickkim0109 committed:
add: 각 모듈 별 Project Path 추가

derrickkim0109 committed:
add: 각 모듈 별 TargetDependency 추가

derrickkim0109 committed:
add: Dependency Plugin 추가

derrickkim0109 committed:
add: Menifest 모듈에 Dependency Plugin 추가

derrickkim0109 committed:
add: SPM 관리를 위한 객체 추가

derrickkim0109 committed:
add: SourceFileList Path 프로퍼티 추가

derrickkim0109 committed:
add: SwiftLint Script 추가

derrickkim0109 committed:
feat: Project 객체 init 메서드 추가

derrickkim0109 committed:
feat: 각 모듈 별 TargetFactory 구현

derrickkim0109 committed:
add: git ignore 파일 추가

derrickkim0109 committed:
add: Tuist version 고정을 위한 파일 추가

derrickkim0109 committed:
add: package version 파일 추가

derrickkim0109 committed:
add: Tuist 각 모듈 Template 추가

derrickkim0109 committed:
feat: App 모듈 초기 설정

- Info.plist 설정
- Package 추가
- AppDelegate, SceneDelegate 추가
- entitlements 파일 추가
- LaunchScreen 추가
- Workspace 파일 추가

derrickkim0109 committed:
add: Feature Onboarding Sample 객체 추가

derrickkim0109 committed:
add: Network Onboarding Sample 객체 추가

derrickkim0109 committed:
add: Service User Sample 객체 추가

derrickkim0109 committed:
add: Shared Sample 추가

- Font Pretendard 추가
- Colors, Font, Images객체 추가
- Utils Sample 객체 추가

derrickkim0109 committed:
feat: make 명령어 추가

- GenerateModule 추가

derrickkim0109 committed:
refactor: Generate Module 네이밍 변경

inwoodev commented on Projects/App/Sources/Application/SceneDelegate.swift:
```diff
@@ -0,0 +1,38 @@
+        }
+
+        window = UIWindow(windowScene: windowScene)
+        let viewController = UIViewController()
+        viewController.view.backgroundColor = .red
```
혹시 이 부분은 작동 확인 하시기 위해 임시로 추가하신걸까요?

derrickkim0109 replied:
네. 맞습니다.
추후에 Router 도입하면서 수정하도록 하겠습니다.

inwoodev commented on Tuist/ProjectDescriptionHelpers/Target+Templates.swift:
```diff
@@ -0,0 +1,301 @@
+        return make(factory: newFactory)
+    }
+
+    static func shared(implements module: ModulePath.Shared, factory: TargetFactory) -> Self {
+        var newFactory = factory
+        newFactory.name = ModulePath.Shared.name + module.rawValue
+        newFactory.sources = .sources
+        
+        if module == .DesignSystem {
+            newFactory.resources = ["Resources/**"]
+            newFactory.product = .staticFramework
+        }
+
+        return make(factory: newFactory)
+    }
```
혹시 designSystem만 staticFramework로 product 설정을 하신 이유도 알 수 있을까요~? :)

derrickkim0109 replied:
Static Framework는 헤더 파일과 라이브러리 코드뿐만 아니라 Resource 파일도 함께 번들링되어 제공되는 것으로 알고 있습니다. `DesignSystem Module`은 Resource 폴더와 코드를 제공하기 위한 모듈로 Static Framework로 설정하는 것이 적합하다고 생각합니다. 

이렇게 설정하면 다른 프로젝트에서 해당 프레임워크를 사용할 때 리소스에 쉽게 접근할 수 있는 장점이 있습니다.

derrickkim0109 committed:
docs: Pull Request Template 작성

derrickkim0109 committed:
docs: commit template 추가

inwoodev committed:
refactor : Tuist의 모듈 설정 수정

- service, network을 모듈에서 제거, core라는 모듈을 통해 Network, LocalDB와 같은 하위 모듈을 관리하도록 설계 수정
- CoreModule이 CLI를 통해 추가될 수 있게 GenerateModule파일 수정
- 수정된 앱 이름을 Tuist 이름 설정에 적용: Lyrics → Feelin

inwoodev commented on Plugins/DependencyPlugin/ProjectDescriptionHelpers/Modules.swift:
```diff
@@ -10,8 +10,7 @@ import ProjectDescription
 
 public enum ModulePath {
     case feature(Feature)
-    case service(Service)
-    case network(Network)
+    case core(Core)
     case shared(Shared)
 }
```
저희가 논의한대로 Service와 Network 모듈 부분을 제거하였고 core path를 추가하였습니다!

inwoodev commented on Plugins/DependencyPlugin/ProjectDescriptionHelpers/Project+Environment.swift:
```diff
@@ -9,9 +9,9 @@ import ProjectDescription
 
 public extension Project {
     enum Environment {
-        public static let appName = "Lyrics"
+        public static let appName = "Feelin"
         public static let deploymentTargets = DeploymentTargets.iOS("15.0")
-        public static let bundleId = "com.project.lyrics"
+        public static let bundleId = "com.project.feelin"
         public static let defaultSettings: Settings = .settings(
             configurations: [
```
팀 논의 후 앱 이름이 Lyrics에서 Feelin으로 변경되었기에 해당 부분을 Project의 Environment세팅에 적용하였습니다.
